proposals/collection-literals.md

```diff
@@ -555,6 +513,107 @@ Not having a *known length* does not prevent any result from being created. Howe
+
+### Natural element type (updated)
+[natural-element-type-updated]: #natural-element-type-updated
+
+Computing the *natural element type* starts with three sets of types and expressions called `dictionary key set`, `dictionary value set`, and `remainder set`.
+
+The `dictionary key/value set` sets will either both be empty or both be non-empty.
+
+Each element of the literal is examined in the following fashion:
+
+* An element `e_n` has its *type* determined.  If that type is some `KeyValuePair<TKey, TValue>`, then `TKey` is added to `dictionary key set` and `TValue` is added to `dictionary value set`.  Otherwise, the `e_n` *expression* is added to `remainder set`.
+
+* An element `..s_n` has its *iteration type* determined.  If that type is some `KeyValuePair<TKey, TValue>`, then `TKey` is added to `dictionary key set` and `TValue` is added to `dictionary value set`. Otherwise, the *iteration type* is added to `remainder set`.
+
+* An element `k_n: v_n` adds the `k_n` and `v_n` *expressions* to `dictionary key set` and `dictionary value set` respectively.
+
+* If the `dictionary key/value set` sets are empty, then there were definitely no `k_n: v_n` elements. In that case, the *fallback case* runs below.
+
+* If `dictionary key/value set` sets are non-empty, then a first round of the *best common type* algorithm in performed on those sets to determine `BCT_Key` and `BCT_Value` respectively.
+
+    * If the first round fails for either set, the *fallback case* runs below.
+
+    * If the first round succeeds for both sets, there is a `KeyValuePair<BCT_Key, BCT_Value>` type produced.  This type is added to `remainder set`.  A second round of the *best common type* algorithm is performed on this set to determine `BCT_Final`.
+
+        * If the second round fails, the *fallback* case runs below.
+        * Otherwise `BCT_Final` is the *natural element type* and the algorithm ends.
+
+* The *fallback case*:
+
+        * All `e_n` *expressions* are added to `remainder set`
+        * All `..s_n` *iteration types* are added to `remainder set`
+        * The *natural element type* is the *best common type* of the `remainder set` and the algorithm ends.
+
+---
+
+* Given:
+
+    ```c#
+    Dictionary<string, object> d1 = ...;
+    Dictionary<object, string> d2 = ...;
+    var d3 = [..d1, ..d2];
+    ```
+
+    The *natural type* of `d3` is `Dictionary<object, object>`.  This is because the `..d1` will have a *iteration type* of `KeyValuePair<string, object>` and `..d2` will have a *iteration type* of `KeyValuePair<object, string>`. These will contribute `{string, object}` to the determination of the `TKey` type and `{object, string}` to the determination of the `TValue` type.  In both cases, the *best common type* of each of these sets is `object`.
+
+* Given:
+
+    ```c#
+    var d = [null: null, "a": "b"];
```

Comment:
This will throw an exception at runtime. Will we naturally just have a warning when the key type violates the nullable contract of the dictionary type?

Reply:
yes.  that should def happen.  good call out and need for tests